File: Core/src/main/java/net/winternetwork/bedwars/core/game/stage/StageManager.java
package net.winternetwork.bedwars.core.game.stage;

import lombok.Getter;
import net.winternetwork.bedwars.api.game.stage.Stage;
import net.winternetwork.bedwars.core.game.GameSettings;

public class StageManager {

    private static StageManager instance;
    private int stageId = 0;
    @Getter
    private Stage actualStage;

    private StageManager() {
        callFirstStage();
    }

    public static StageManager getInstance() {
        return instance == null ? instance = new StageManager() : instance;
    }

    private void callFirstStage() {
        actualStage = GameSettings.STAGE_ARRAY[0];
        actualStage.onStageJoin();
    }

    public void callNextStage() {
        final Stage stage = actualStage;

        if (stage != null) {
            stage.onStageExit();
        }

        actualStage = GameSettings.STAGE_ARRAY[++stageId];
        actualStage.onStageJoin();
    }

    public void callPreviousStage() {
        actualStage = GameSettings.STAGE_ARRAY[--stageId];

        actualStage.onStageJoin();
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/modules/stage/StageManager.kt
package net.winternetwork.bedwars.modules.stage

import net.winternetwork.bedwars.api.game.stage.Stage
import net.winternetwork.bedwars.game.settings.GameSettings

class StageManager(
        private val stages: List<Stage>
) {
    lateinit var actualStage: Stage
        private set

    private var i = 0

    init {
        if (GameSettings.canStart) {
            actualStage = stages[0]
            actualStage.onStageJoin()
        }
    }

    fun next() {
        val stage = actualStage

        stage.onStageExit()

        actualStage = stages[++i]
        actualStage.onStageJoin()
    }

    fun previous() {
        actualStage = stages[--i]
        actualStage.onStageJoin()
    }
}